### ISSUE TITLE
[A11] Unable to access exported bookmarks

### ISSUE THREAD

**Issue by DefiCzech** [bromite/bromite#831]
### Bromite version

Version: `87.0.4280.81`
Arch: `arm64`
Android version: `11.0`
Device model: `RP1A.201005.006`

### Is this bug about the SystemWebView?
No
### Is the bug reproducible with latest version?
Yes
### Can the bug be reproduced with corresponding Chromium version?
No

### Describe the bug
On my device Google Pixel with Android 11 the exported bookmarks file `bookmarks.html` is saved (and opened in Bromite) on this location `file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html`, but I have no access to this location outside Bromite. So I am unable to copy this file and backup on other device.

### Steps to reproduce the bug
1. open bromite menu (right top three dots)
2. click to bookmarks
3. click to export bookmarks icon (down arrow)
4. bookmarks.html is create and opened, but unable to save to another location and inaccessible on current location.

### Expected behavior
- It would be nice save this file to same location as other downloaded files from internet `filte:///storage/emulated/0/Download/bookmarks.html`.
- Or show popup window, where user can choose the location where to save exported bookmarks
- Or have the possibility to save exported and opened bookmarks.html to another accessible location

**Comment by Poopooracoocoo** [comment]
I think this is an issue that only occurs on Android 11 and higher because of scoped storage btw.

**Comment by TheProUser** [comment]
Well, there is a workaround for that. 

1) First, export all the bookmarks

2) Paste this file link 

`view-source:file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html` 

in a new bromite tab. If your phone is slow, it might take time to load the source. Scroll to the bottom of the page to let it fully load. 

3) Copy everything from the page. **While doing that, don't press on a link.** To correctly do that, zoom in the page, long press on a **text**, press on the option **Select all** and then on **Copy" from the popup. 

4) Create a **file** with `.html` extension in your desirable directory from a good file manager like "Mixplorer" or "Solid File Manager". Open it. It will be empty. Paste what you copied before in it. 

Viola!

**Comment by TheProUser** [comment]
Well, there is a workaround for that. 

1) First, export all the bookmarks

2) Paste this file link 

`view-source:file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html` 

Or this one 

`view-source:file:///sdcard/Android/data/org.bromite.bromite/files/Download/bookmarks.html`

in a new bromite tab. If your phone is slow, it might take time to load the source. Scroll to the bottom of the page to let it fully load. 

3) Copy everything from the page. **While doing that, don't press on a link.** To correctly do that, zoom in the page, long press on a **text**, press on the option **Select all** and then on **Copy** from the popup. 

4) Create a **file** with `.html` extension in your desirable directory from a good file manager like "MiXplorer" or "Solid Explorer" etc. Open it. It will be empty. Paste what you copied before in it. Save and exit.

Viola!

**Comment by DefiCzech** [comment]
Yes this workaround is doable but not for less experienced user, who needs to backup bookmarks repeatedly. Too much steps to remember. So for me it is bad user experience.

**Comment by csagan5** [comment]
@thestinger do you think there is a solution to this? e.g. what are we doing wrong here? https://github.com/bromite/bromite/blob/87.0.4280.81/build/patches/Add-bookmark-import-export-actions.patch#L499

**Comment by csagan5** [comment]
@DefiCzech you need to submit this with the issue template for bugs, not features.

**Comment by DefiCzech** [comment]
@csagan5 template changed

**Comment by uazo** [comment]
here maybe some info

https://source.chromium.org/chromium/chromium/src/+/master:base/android/java/src/org/chromium/base/PathUtils.java;l=206

**Comment by csagan5** [comment]
> here maybe some info

@uazo thanks; it is then working as intended on Android 11.

**Comment by thestinger** [comment]
It *could* use SAF to ask the user for a directory but it would have to be rewritten to support that.

**Comment by csagan5** [comment]
In such case, on A11, the permission would be granted because the user dialog interaction happened?

**Comment by thestinger** [comment]
See https://developer.android.com/guide/topics/providers/document-provider. This is the generic way that access to shared files and directories is supposed to be done. It doesn't use permissions.

It has been available since Android 4.4 and gained support for directories in 5.0. Apps have not needed to request a coarse-grained permission to access storage for years but chose to keep doing things in the way that was easiest for them.

The intention of scoped storage is not really that apps stick to their scoped app directories. The intention is that they finally move to the system introduced for user control and user consent over file / directory access. Chromium probably doesn't want to overhaul their code to cover niche use cases. This is what they *should* have done to support choosing the download location on Android 10+.

**Comment by csagan5** [comment]
> Chromium probably doesn't want to overhaul their code to cover niche use cases. This is what they _should_ have done to support choosing the download location on Android 10+.

Damn...so now here we would have to use the same code we use for the download location choice option.

**Comment by wchen342** [comment]
I have a modificationed version of the file selection dialog using `ACTION_OPEN_DOCUMENT_TREE` in [this patch](https://github.com/ungoogled-software/ungoogled-chromium-android/blob/extensions/patches/Extensions/allow-select-folder.patch). It does not always work when used for extensions (partially because the reading is done in native codes rather than Java), but I think you can get an idea how it works.

**Comment by uazo** [comment]
I tried with the A11 emulator and now it ask for permission.
@DefiCzech could you check?

**Comment by DefiCzech** [comment]
@uazo I have Bromite version `88.0.4324.149`. After clicking download bookmarks. I ask for permission to save to my storage. After granting access to my storage, the bookmarks.hmtl is still saved to `file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html` from where I am unable to backup the file.

**Comment by csagan5** [comment]
Fixed in `88.0.4324.185`; @DefiCzech can you please confirm?

**Comment by DefiCzech** [comment]
@csagan5 sadly no. I updated Bromite to version (I tried reinstalling also) `88.0.4324.185` from fdroid. After clicking download bookmarks and allowing access to storage, the bookmarks.html is automatically saved to `file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html` location, no _choose where to save_ dialog or saved to location where I have access to.

**Comment by uazo** [comment]
I think we have two alternatives:
1) add @wchen342 's patch
2) considering that we are now asking for write permissions, add the possibility to save directly in the "ExternalStoragePublicDirectory" also for A11

(1) would allow us to use directory selection in any patch in the future, (2) obviously is quicker to develop.

**Comment by csagan5** [comment]
@uazo two questions:
* do the changes from #896 (or part of them) need to be reverted? Since they were not effective
* are the 2 approaches equivalent for the user? I would be inclined to go for (1) so that there is less need to rewrite it in future relases (A11 will catch up fast as soon as newer Android phones come out)

**Comment by wchen342** [comment]
I need to point out that my patch uses `ACTION_OPEN_DOCUMENT_TREE` and has extra codes because in my case (developer mode loading) the file paths need to be parsed from URIs for native codes to read the files.
I think for bookmark import/export you do not need to do that because (1) it operates on one file at a time and (2) you can do the file operations using pure java so you can use the URIs directly.

**Comment by uazo** [comment]
> do the changes from #896 (or part of them) need to be reverted?

no, I need those changes for the (1)

> are the 2 approaches equivalent for the user?

unfortunately I don't think it's the same thing. with the (1) it seems that the user cannot specify the download folder:
```
Access restrictions
On Android 11 (API level 30) and higher, you cannot use the ACTION_OPEN_DOCUMENT_TREE 
intent action to request access to the following directories:

- The root directory of the internal storage volume.
- The root directory of each SD card volume that the device manufacturer considers to be reliable, 
regardless of whether the card is emulated or removable. A reliable volume is one that an app can
successfully access most of the time.
- The Download directory.
```
from https://developer.android.com/training/data-storage/shared/documents-files

having no direct experience, I would try to do both (1) and (2) and then we decide what to leave and what to delete.
the strange thing is that from the emulator I have access to `file:///storage/emulated/0/Android/data/org.bromite.bromite/files/Download/bookmarks.html` with the default A11 filemanager, maybe it is a bug of the user's android distribution

@wchen342 could you tell me why you say it doesn't work every now and then? eventually it would be enough for me just to know where you use it in the code, simply because "developer mode loading" I don't understand what that means

**Comment by DefiCzech** [comment]
Maybe it is totaly misleading, but in NewPipe (https://github.com/TeamNewPipe/NewPipe) I am able choose location where export data (but it is not using system dialog). In Threema (https://github.com/threema-ch/threema-android) I can choose location (system dialog) where to create my ceversations backup. In NetGuard (https://github.com/M66B/NetGuard) I can choose location (system dialog) where to create my settings backup.

**Comment by wchen342** [comment]
> choose location where export data (but it is not using system dialog)

The NewPipe one is a reskinned system dialog I think.

> it doesn't work every now and then

I was referring to the part that parses uris to absolute paths (the `private static final class FileUtil` part). It is because of the name of primary volume are inconsistent between devices and some functions used will potentially be greylisted. As I said it is for native code to read files so you won't need it here.

> because "developer mode loading" I don't understand what that means

That was just the extension developer mode. The reason I added folder picking ability is because in normal chromium if you want to load extensions from disk you need to enter developer mode and select a folder.

**Comment by uazo** [comment]
since, as I said, the (2) is faster to do (for me :), here the resulting apk.

arm64: https://www.mediafire.com/file/v2qdg5gyv24iytz/ChromePublic.apk/file

if @DefiCzech can test it... The button is "(WIP) export to downloads"

**Comment by DefiCzech** [comment]
Save bookmarks to `file:///storage/emulated/0/Download/bookmarks.html` so it works. Amazing.

**Comment by uazo** [comment]
> Saved bookmarks to file:///storage/emulated/0/Download/bookmarks.html so it works. 

ok, perfect, i was just curious. I am compiling the other version ..

**Comment by uazo** [comment]
unfortunately I am having problems with that code, certainly due to my inexperience.

```
02-19 08:18:20.952  9149  9149 D PickerActionHandler: onFinished() [content://com.android.externalstorage.documents/tree/home%3A]
02-19 08:18:20.988 11558 11622 I cr_SelectFileDialog.java: getVolumeIdFromTreeUri: docId: home:
02-19 08:18:20.988 11558 11622 I cr_SelectFileDialog.java: getVolumePath: volumeId: home
02-19 08:18:20.991 11558 11622 I cr_SelectFileDialog.java: getVolumePath: length: 1
02-19 08:18:20.992 11558 11622 I cr_SelectFileDialog.java: getVolumePath: uuid: null
02-19 08:18:20.992 11558 11622 I cr_SelectFileDialog.java: getVolumePath: primary: true
02-19 08:18:20.992 11558 11622 I cr_SelectFileDialog.java: volumePath: NULL
02-19 08:18:20.992 11558 11622 I cr_SelectFileDialog.java: mUris: file:///, path: /
02-19 08:18:20.992 11558 11622 I cr_SelectFileDialog.java: mFilePaths1: /
02-19 08:18:20.993 11558 11622 I cr_SelectFileDialog.java: displayNames1: /
```

I try a little longer, in case I can't, I go around it.
I have read that there are various types of `content://`, including "virtual" files which if I am not mistaken are, for example, the files exposed by programs such as dropbox/owncloud. seems to me that the chromium code to be able to handle them even natively.

so the idea could be to allow not the selection of the folder, but directly the name of the file (using Save As): theoretically so you could save directly to the cloud ..

If this works, I keep the name in a preference so that next time there is no need to reselect it.
If I can't, I use the download patch code!

**Comment by wchen342** [comment]
From the log it seems the dialog didn't get the uri at all.

From the uri you seems to be selecting SD card? I think that is not allowed by Android isn't it? Normal packages are only allowed to rw internal storage since A9.

**Comment by wchen342** [comment]
Still in the case of bookmark it is much better to use `ACTION_OPEN_DOCUMENT` instead. You do not need access to a folder anyway unless you want to save multiple files at a time. In the case you need to run native code you can just create a JNI bridge and get the result from C++. Since the output is string it shouldn't be hard.

**Comment by uazo** [comment]
@csagan5 i ended up in an android bug: https://issuetracker.google.com/issues/37136466

for which:

1) export bookmarks -> open file dialog with default "bookmarks.html" -> the file does not exist -> the returned file is "bookmarks.html"

2) export bookmarks -> open file dialog with default "bookmarks.html" but the file already exists -> the returned file is "bookmarks.html (1)"

3) export bookmarks -> open file dialog with default "bookmarks.html" but the file already exists and the user selects it -> asks for overwriting (only A9 +) -> the returned file is "bookmarks.html".

rather disappointing.

i think i will have to change the export to:
1) export bookmarks -> check last file selected exists -> ask for overwrite -> no: open dialog

other ideas?

**Comment by uazo** [comment]
I am slowly understanding why some things are not active in chromium...

For example SAF doesn't seem to be supported by dropbox, while in owncloud I have several problems (see https://github.com/owncloud/android/issues/3110)

let's see if they help me understand something, in the hope that it is not some limit imposed by android for browsers

**Comment by csagan5** [comment]
> i ended up in an android bug: https://issuetracker.google.com/issues/37136466

It is marked as fixed; in case of (2) can we reject the file and take back the user to the dialog?

> For example SAF doesn't seem to be supported by dropbox, while in owncloud I have several problems

This is unrelated to this bug.

@uazo I do not think we need to reinvent the wheel. We already have all the functionality on the prompt for regular downloads, which is enabled by default. I just checked on A10 and A11; on A11 it does not show the SD card option, but it still shows a little dialog and if file exists it shows `filename (1).ext`.

Can we copy it? If that one works mostly via Java, we can do the corresponding UI part also via Java?

There are some discussions about the download prompt in A11 in https://github.com/GrapheneOS/Vanadium/issues/17, https://github.com/GrapheneOS/Vanadium/issues/90 and #485
If that is working, can we copy it?

**Comment by uazo** [comment]
> It is marked as fixed; 

yes, I saw, but not only for who can upgrade the system?
maybe i still don't understand how android ecosystem works, let me understand, can i patch android libraries in chromium too?
let me explain better, if I patch a file as it does here https://chromium-review.googlesource.com/c/chromium/src/+/2057806/7/third_party/android_provider/java/src/org/chromium/third_party/android/provider/MediaStoreUtils.java in chromium repository, in your experience, I carry the changes in chromium apk?

> in case of (2) can we reject the file and take back the user to the dialog?

I think so, I just need to check the extension, but it's really bad in my opinion

>  I do not think we need to reinvent the wheel.

Why do you say that? after all I'm using something that should be standard. if I point to a local folder it works perfectly: but since the filedialog allows the selection of all registered providers, I went ahead with any comments from users regarding possible bugs that are not caused by my code.

> Can we copy it?

that code doesn't uses saf, but points the file directly the old way. in that way I wouldn't have fixed anything: rather to quickly fix this bug just ask for permissions and point to the correct folder, as see https://github.com/bromite/bromite/issues/831#issuecomment-781455177 if you want I release the patch with those changes, that it seemed to me at no cost to be able to choose the destination, but I was wrong

**Comment by csagan5** [comment]
> in chromium repository, in your experience, I carry the changes in chromium apk?

No, I exclude that. The Android code present there for Chromium is like "headers" in the C++ world.

> > in case of (2) can we reject the file and take back the user to the dialog?
> 
> I think so, I just need to check the extension, but it's really bad in my opinion

Can you please elaborate?

> Why do you say that?

Because I tested with A10 and A11 and found a save dialog that works for both (the Download one). 

> that code doesn't uses saf, but points the file directly the old way. in that way I wouldn't have fixed anything

I do not follow. I tested all the scenarios. Can we make a list of scenarios to test? Maybe I missed something. The tests I did behaved correctly both for existing and non-existing files, and with both A10 and A11.

> it seemed to me at no cost to be able to choose the destination, but I was wrong

Yes, this is possible also with the Downloads file UI.

**Comment by uazo** [comment]
> No, I exclude that. The Android code present there for Chromium is like "headers" in the C++ world.

it depends, i can modify androidx from what i saw

> Can you please elaborate?

because there is no way to know if it is the user's will or not, the framework does not in any way make it possible for you to know.

saf is actually well done, but according to what I have been able to see, it rests on a concept that is misleading for the user: it wants to be a pure virtual file system (and it is) but does not oblige the implementers of providers to the concept of uniqueness of the file (document for saf), while the user expects that a file with the same name (display-name for saf) to be the same in the same folder.

In fact, for saf, one thing is the display name, one thing is the name of the corresponding file in the storage, which may have no affinity because it is left to the provider implementer to choose when, how, where and whether to save the file and, again for saf, different files can have the same display-name.

this is evident when a provider must be virtual and physical at the same time, like those based on the cloud, because it is virtual on android, but physical on other devices, therefore related to the same concept display-name same filename.

they probably noticed it later, and instead of binding this logic of uniqueness, they added that weird behavior to the pickerui (add (1..x) to the display name, which I still not found in the android codebase, but must be somewhere), instead of definitively impose that the display-name should be unique.

to this is added the fact that the choice of when to generate what saf calls "documentid" and whether it should be unique for new files is left to the provider implementer: there is nothing that forces it and therefore makes its own choices and so you fall into the bug indicated in https://github.com/owncloud/android/issues/3110 which I can not avoid in any way, now i try with `getClipData()`

>> that code doesn't uses saf, but points the file directly the old way. in that way I wouldn't have fixed anything

> Yes, this is possible also with the Downloads file UI.

is it possible to select a directory like gdrive or owncloud as the destination? if so, where is it in the code that i can't find it?

**Comment by csagan5** [comment]
> it depends, i can modify androidx from what i saw

https://android-developers.googleblog.com/2018/05/hello-world-androidx.html

AndroidX is support libraries; you cannot change the OS (Android) source.



> > > that code doesn't uses saf, but points the file directly the old way. in that way I wouldn't have fixed anything
> 
> > Yes, this is possible also with the Downloads file UI.
> 
> is it possible to select a directory like gdrive or owncloud as the destination? if so, where is it in the code that i can't find it?

What I am saying is that for the exported bookmarks having the same UI that currently is in the Downloads will suffice. Nothing else is necessary; in fact, I will not accept patches to fix this issue that are any different than that.

**Comment by uazo** [comment]
> I will not accept patches to fix this issue that are any different than that

it's a real shame, however, consider the security aspect, to allow A11 to save on external downloads, it is necessary to request the permission of "WRITE_EXTERNAL_STORAGE", with saf it is not necessary

**Comment by csagan5** [comment]
> it's a real shame, however, consider the security aspect, to allow A11 to save on external downloads, it is necessary to request the permission of "WRITE_EXTERNAL_STORAGE", with saf it is not necessary

The bugfix here is to have the same UI as the downloads; if you want to research and experiment for a SAF solution it should be separate.

This issue is not "let's add cloud/SAF support to the saved bookmarks", by sticking closely to upstream we can deliver a fix that does not require too many changes/maintenance and the feature (for SAF) can be developed separately.

**Comment by uazo** [comment]
> The bugfix here is to have the same UI as the downloads; if you want to research and experiment for a SAF solution it should be separate.

unfortunately it was not possible for me to separate the two, because, based on the tests done on the emulator, I have no access to external storage in any way except using saf.
so the patch exposes both possibilities: by default using the interface proposed by the download, and under Q and later (or with the activation of the flag in chrome://flags) using the fileselector and saf.

here the result:

**Without SAF;**
![without_saf](https://user-images.githubusercontent.com/29201891/111341438-cf2e1280-8679-11eb-8559-386f059ab8fb.gif)

**With SAF:**
![with_saf](https://user-images.githubusercontent.com/29201891/111341477-d81ee400-8679-11eb-90a9-323051f469c5.gif)

https://github.com/bromite/bromite/commit/2a2a3d89ed9d5f5453c4e9892b144f88ebdd7dd4
here you find the patch, tell me if it is appropriate to propose a pull request

> by sticking closely to upstream we can deliver a fix that does not require too many changes/maintenance and the feature (for SAF) 

just to let you know, I made more changes trying to reuse the download interface than using saf ...

**Comment by csagan5** [comment]
Where is the version without SAF?

**Comment by uazo** [comment]
here you find the method that reuses the download ui:
https://github.com/bromite/bromite/blob/2a2a3d89ed9d5f5453c4e9892b144f88ebdd7dd4/build/patches/Add-bookmark-import-export-actions.patch#L463

is called from here:
https://github.com/bromite/bromite/blob/2a2a3d89ed9d5f5453c4e9892b144f88ebdd7dd4/build/patches/Add-bookmark-import-export-actions.patch#L349-L355

if you see it is under a currently disabled flag except for Q+ since the inability to reuse that interface due to the limitations imposed in android. I could have changed that too, but the intervention would have been invasive and useless and in R it would never work.

saf code is only in 
https://github.com/bromite/bromite/blob/2a2a3d89ed9d5f5453c4e9892b144f88ebdd7dd4/build/patches/Add-bookmark-import-export-actions.patch#L358-L358
and the new methods in `ContentUriUtils.java` and `content_uri_utils.cc`.
everything else is the old patch and the changes made to be able to access the download interface from a child class (exposing the variables as protected, so as not to touch the current code)

**Comment by csagan5** [comment]
So it is a single patch with a flag to switch between the two?

**Comment by uazo** [comment]
yes, it is.

**Comment by csagan5** [comment]
@uazo some things I noticed:
* when using SAF if the user clicks on an already existing file then a new prompt to overwrite the file appears; this makes the first prompt a duplicate
* if the user makes manual changes to the filename and clicks on Save then no overwrite prompt is displayed
* next time user makes an export the last filename is used, not `bookmarks.html`

**Comment by csagan5** [comment]
We can continue with this but I would make SAF enabled by default and create a new issue to have the Downloads one changed as well. What do you think @uazo?

**Comment by uazo** [comment]
> when using SAF if the user clicks on an already existing file then a new prompt to overwrite the file appears; this makes the first prompt a duplicate

no chance to check/change this, blame the android developers :)!

> if the user makes manual changes to the filename and clicks on Save then no overwrite prompt is displayed

like above

> next time user makes an export the last filename is used, not bookmarks.html

yes, it is intended.

> to have the Downloads one changed as well. What do you think @uazo?

I don't know, that change is very very very invasive, I wanted to do it, but I gave up, too many modified files... to be put in very low priority.

**Comment by csagan5** [comment]
Fixed in `89.0.4389.100`.
